### PR TITLE
JSON: Extract register information

### PR DESCRIPTION
**Pull request by AbhinavMir**
(currently a WIP, still figuring things out)

Potential fix for #10

### CODE REVIEW

**Comment by AbhinavMir** [comment]
```json
....

{
  "mnemonic": "zip",
  "name": "TBD",
  "operands": [ { "name": "rs1", "type": "regidx" },{ "name": "rd", "type": "regidx" } ],
  "syntax": "rd,rs1",
  "format": "TBD",
  "fields": [ { "field": "0b000010001111", "size": 12 }, { "field": "rs1", "size": 5 }, { "field": "0b001", "size": 3 }, { "field": "rd", "size": 5 }, { "field": "0b0010011", "size": 7 } ],
  "extensions": [ "Zbkb" ],
  "function": "{\n  assert(sizeof(xlen) == 32);\n  let rs1_val = X(rs1);\n  result : xlenbits = zeros();\n  foreach (i from 0 to (sizeof(xlen_bytes)*4 - 1)) {\n    result[i*2] = rs1_val[i];\n    result[i*2 + 1] = rs1_val[i + sizeof(xlen_bytes)*4];\n  };\n  X(rd) = result;\n  RETIRE_SUCCESS\n}",
  "description": "TBD"
}
  ],
  "registers": 
[{
  "name": "x31",
  "type": "regtype",
  "description": "General Purpose Register"
},
{
  "name": "x5",
  "type": "regtype",
  "description": "General Purpose Register"
},
{
  "name": "x1",
  "type": "regtype",
  "description": "General Purpose Register"
},
{
  "name": "x2",
  "type": "regtype",
  "description": "General Purpose Register"
},
{
  "name": "x20",
  "type": "regtype",
  "description": "General Purpose Register"
},
{
  "name": "x21",
  "type": "regtype",
  "description": "General Purpose Register"
},
{
  "name": "x15",
  "type": "regtype",
  "description": "General Purpose Register"
},
{
  "name": "PC",
  "type": "xlenbits",
  "description": "Program Counter"
},
{
  "name": "x16",
  "type": "regtype",
  "description": "General Purpose Register"
},
{
  "name": "x9",
  "type": "regtype",
  "description": "General Purpose Register"
},
{
  "name": "instbits",
  "type": "xlenbits",
  "description": "Instruction Register"
},
{
  "name": "x19",
  "type": "regtype",
  "description": "General Purpose Register"
},
{
  "name": "x10",
  "type": "regtype",
  "description": "General Purpose Register"
},
{
  "name": "x11",
  "type": "regtype",
  "description": "General Purpose Register"
},
{
  "name": "x23",
  "type": "regtype",
  "description": "General Purpose Register"
},
{
  "name": "x27",
  "type": "regtype",
  "description": "General Purpose Register"
},
{
  "name": "x12",
  "type": "regtype",
  "description": "General Purpose Register"
},
{
  "name": "x29",
  "type": "regtype",
  "description": "General Purpose Register"
},
{
  "name": "x28",
  "type": "regtype",
  "description": "General Purpose Register"
},
{
  "name": "x8",
  "type": "regtype",
  "description": "General Purpose Register"
},
{
  "name": "x17",
  "type": "regtype",
  "description": "General Purpose Register"
},
{
  "name": "x6",
  "type": "regtype",
  "description": "General Purpose Register"
},
{
  "name": "nextPC",
  "type": "xlenbits",
  "description": "Next Program Counter"
},
{
  "name": "x25",
  "type": "regtype",
  "description": "General Purpose Register"
},
{
  "name": "x26",
  "type": "regtype",
  "description": "General Purpose Register"
},
{
  "name": "x18",
  "type": "regtype",
  "description": "General Purpose Register"
},
{
  "name": "x30",
  "type": "regtype",
  "description": "General Purpose Register"
},
{
  "name": "x4",
  "type": "regtype",
  "description": "General Purpose Register"
},
{
  "name": "x22",
  "type": "regtype",
  "description": "General Purpose Register"
},
{
  "name": "x3",
  "type": "regtype",
  "description": "General Purpose Register"
},
{
  "name": "x7",
  "type": "regtype",
  "description": "General Purpose Register"
},
{
  "name": "x24",
  "type": "regtype",
  "description": "General Purpose Register"
},
{
  "name": "x14",
  "type": "regtype",
  "description": "General Purpose Register"
},
{
  "name": "x13",
  "type": "regtype",
  "description": "General Purpose Register"
}]
,
  "formats": [
"I",
"R",
"TBD"
  ],
  "extensions": [
"A",
"C",
"D",
"F",
"M",
"V",
"Zba",
"Zbb",
"Zbc",
"Zbkb",
"Zbkc",
"Zbkx",
"Zbs",
"Zfa",
"Zfh",
"Zicond",
"Zknd",
"Zkne",
"Zknh",
"Zksed",
"Zksh"
  ]
}
```


Does this look good?

**Comment by ThinkOpenly** [comment]
The look it spot on. However, the origin of the emitted data is not the Sail code. You are generating the data (using very nice OCaml!). So let's back up a bit...

The goal is to extract everything from the Sail code. The basic flow is that the main Sail parser reads all the Sail code, converts it into an AST, then passes it to the code in `src/sail_json_backend/json.ml`. Here, starting in the function `defs`, we walk through the AST and store the pieces that we need in hash tables. This is where the register extraction should be happening. All of the registers are defined like:
```
register x1 : regtype
```

That `register` keyword is translated into _something_ in the AST. We need to figure out what that is, and the context in which we'll encounter it while walking through the AST. It looks a lot like a top-level element in the Sail code, so that makes things a bit simpler.

Diving a bit deeper...

I'm not an expert in the AST data structures, so let's try something rudimentary:
```
$ grep -i reg src/lib/parse_ast.ml
  (* index specification, for bitfields in register types *)
  | TD_bitfield of id * atyp * (id * index_range) list (* register mutable bitfield type definition *)
type dec_spec_aux = (* Register declarations *)
  | DEC_reg of atyp * id * exp option
  | DEF_register of dec_spec (* register declaration *)
```

Those last two look interesting, and in particular, that last one. If you look at the top of the `defs` function in `json.ml`, you'll see that we're iterating over the list of `defs` (parameter same name as the function, I see... hmm... maybe I could've chosen a better set of names here, anyway...), and they are trying to match against `DEF_*` enums/records. I think we're on to something! Try adding a new matching case here for `DEF_register` and maybe just print something to know that you got a hit:
```
  DEF_register (_) -> print_endline "DEF_register!"
```

Presuming that works, we need to dive into what is attached to the `DEF_register` record. From the `grep` output above, it looks like a `dec_spec`. Going back to `parse_ast.ml`, we see:
```
type dec_spec_aux = (* Register declarations *)
  | DEC_reg of atyp * id * exp option
[...]
type dec_spec = DEC_aux of dec_spec_aux * l
```

At this point, we can enhance our match case above:
```
  DEF_register (DEC_aux (DEC_reg (a, i, e), _)) -> print_endline ("DEF_register: " ^ (string_of_id i))
```

And maybe (hopefully!) we see some more detailed information! Keep scratching at the data until you get that there is a register called "x1" of type "regtype", and stuff this information in a hash table.

Notice that we don't have any human-readable type information, like "General Purpose Register". If we want this in our JSON, we'll need to figure out a way to add it to the Sail code. (That can be a later phase, of course.)

**Comment by AbhinavMir** [comment]
Thanks for the feedback! I was going to bring it up tmrw, but I was sort of just testing where I am going with this and form helpers around it (`json_of_registers`, `let registers ...` etc) and make sure it builds for me instead of erroring out. Another thing was, does the register have an expected output in the form of a description (which is just a GPR placeholder for now)?

Just going one step at a time 😅

**Comment by AbhinavMir** [comment]
@ThinkOpenly  I finally figured out the problems in my previous approach of not using AST and now I use the `DEF_register` as you suggested but it generates register entries for all sorts of registers. I considered filtering out registers of type `regtype` only but didn't see [latex.ml](https://github.com/AbhinavMir/sail/blob/cf6b39a78f9e99c05bbd6d558058eb8d5cd121b2/src/sail_latex_backend/latex.ml#L536-L539) doing so. I also see references to a few configuration registers in the `generated_definitions/ast` - not sure where to go from here.

**Comment by ThinkOpenly** [comment]
W00t! This is fantastic! Good work!

> it generates register entries for all sorts of registers.

Yep, this is expected. Indeed many of the other registers are pretty basic, like floating-point:
```
model/riscv_fdext_regs.sail:register f0  : fregtype
```

and vector:
```
model/riscv_vext_regs.sail:register vr0 : vregtype
```

and CSR:
```
model/riscv_sys_regs.sail:register misa : Misa
```

> I considered filtering out registers [...]

No need. I confess I don't know the meaning/purpose of all of the "registers", but I'd rather have a superset including the essential registers at this point.

> does the register have an expected output in the form of a description (which is just a GPR placeholder for now)?

In summary, what you have now has a lot of value. If you want to get it merged, I can do a detailed review now. Things that could be done now, or after merging:
- Sorting by type, then name (that way all of the floating point registers are together, in order, for example)
- Adding the descriptions. This might be a bit of work, and how to do it is subjective. One way is to add an attribute to the _type_, and then use that for each register, _or_ maybe create yet another top-level entry in the JSON for "types" that might look something like:
```
"types": [
  {
    "regtype": {
      "description": "general purpose register",
      "size": "xlen"
    },
    "Counteren": {
      "description": "counter enable",
      "size": "32",
      "fields": [
        { "name": "HPM", "size": "29" },
        { "name": "IR", "size": "1" },
[...]
```

**Comment by AbhinavMir** [comment]
Thanks @ThinkOpenly ! I was thinking of doing sorting now and separating descriptions into a separate issue - since I imagine it'll be some work. Are all registers in sail_riscv defined with their descriptions or will it be something I'll be defining in `json.ml` per register type. I personally think attribute to the type looks cleaner - do you sway either way on this topic or are you tending neutral on this?

Also I'll squash commits in into new commits by the end.

Cheers.

**Comment by ThinkOpenly** [comment]
> I was thinking of doing sorting now and separating descriptions into a separate issue - since I imagine it'll be some work.

Agreed.

> Are all registers in sail_riscv defined with their descriptions or will it be something I'll be defining in `json.ml` per register type.

I don't think any registers currently have anything that helps a human identify them, except for comments, which are far too unwieldy. So, we'll need to add them, at least for the registers we care about, which is subjective at this point.

> I personally think attribute to the type looks cleaner

I definitely agree, but it's a bit more work, but also a bit more value.

I don't know if I explained attributes. Here's an example of realization:
```
mapping itype_mnemonic : iop <-> string = {
  $[name add immediate]
    RISCV_ADDI  <-> "addi",
  $[name set less than immediate]
    RISCV_SLTI  <-> "slti",
[...]
}
```
An example on the parsing side can be found in `json.ml` in `parse_mapcl`. In the AST, the attribute appears in an "annotation" tied to the `MCL_aux` type (the part we usually ignore). Near the bottom of the function is where we add the discovered attribute information into the `names` hash table.

**Comment by AbhinavMir** [comment]
Looking into `parse_mapcl`! For the first bit, wondering if I'd have to define register descriptions in json.ml or in the sail-riscv repostiory?

**Comment by ThinkOpenly** [comment]
> define register descriptions in json.ml or in the sail-riscv repostiory?

Definitely in the sail-riscv repo. We want to keep "apriori" information out of `json.ml` if at all possible. The sail code should be sort of a "black box", with everything essential encoded in the AST, but that's a very long-term goal. The "identity functions" in `json.ml` now are an abomination. It would be nice to eventually get rid of that. Even "extension" is a bit of a hack, but maybe a reasonable one.

**Comment by ThinkOpenly** [comment]
I imagine applying an attribute to the types would look something like:
```
$[name general purpose register]
type regtype = xlenbits
```

**Comment by AbhinavMir** [comment]
Okay, got it! Also, I finished the sorting bit too, should I squash the commits into one PR and mark this ready for review?

**Comment by ThinkOpenly** [comment]
> I finished the sorting bit too, should I squash the commits into one PR and mark this ready for review?

Awesome! Yeah, go for it!

File: src/sail_json_backend/json.ml [hunk 1]
```diff
@@ -147,6 +148,7 @@ let rec string_list_of_mpat x = match x with
       List.concat (List.map string_list_of_mpat mpl)
   | _ -> assert false
 
+
```
Comment (ThinkOpenly) [inline src/sail_json_backend/json.ml]: Minimize extraneous changes, especially in whitespace. It makes the code easier to review if the only changes are code changes. Here, a blank line was added.

File: src/sail_json_backend/json.ml [hunk 2]
```diff
@@ -202,6 +204,24 @@ let add_assembly app_id p =
       Hashtbl.add assembly (string_of_id app_id) x
     end
 
+
+
```
Comment (ThinkOpenly) [inline src/sail_json_backend/json.ml]: 2 blank lines added.

File: src/sail_json_backend/json.ml [hunk 3]
```diff
@@ -202,6 +204,24 @@ let add_assembly app_id p =
...
+let json_of_registers () =
+  let regs = Hashtbl.fold (fun name regtype accum ->
+    (name, regtype) :: accum
+  ) registers [] in
+  let sorted_regs = List.sort (fun (_, t1) (_, t2) -> String.compare t1 t2) regs in
```
Comment (ThinkOpenly) [inline src/sail_json_backend/json.ml]: This sorts by regtype only. Could you add a secondary sort, by regname? I think without this, you may get registers in odd orders, no? Like "x0", "x4", "x17", "x1", ...?

File: src/sail_json_backend/json.ml [hunk 4]
```diff
@@ -202,6 +204,24 @@ let add_assembly app_id p =
...
+    "\n  {\n" ^
+    "    \"name\": \"" ^ name ^ "\",\n" ^
+    "    \"type\": \"" ^ regtype ^ "\"\n  },"
+  ) sorted_regs in
+  let reg_list = String.concat "" json_regs in
+  let final_reg_list = if reg_list = "" then "" else String.sub reg_list 0 (String.length reg_list - 1) in
+  "[ " ^ final_reg_list ^ "\n]"
+    
+
+
```
Comment (ThinkOpenly) [inline src/sail_json_backend/json.ml]: Two extra blank lines.
The convention I've used so far is just a single blank line between functions.

File: src/sail_json_backend/json.ml [hunk 5]
```diff
@@ -302,6 +322,7 @@ let parse_type_union i ucl =
       end;
       debug_print ")"
 
+
```
Comment (ThinkOpenly) [inline src/sail_json_backend/json.ml]: blank line added.

File: src/sail_json_backend/json.ml [hunk 6]
```diff
@@ -432,7 +453,7 @@ let json_of_syntax k =
     else (remove_identity_funcs s)) (List.tl (Hashtbl.find assembly_clean k)) in
       if List.length l > 0 then (String.concat "" l)
       else ""
-
+
```
Comment (ThinkOpenly) [inline src/sail_json_backend/json.ml]: blank line (with spaces) added.

File: src/sail_json_backend/json.ml [hunk 7]
```diff
@@ -600,9 +621,12 @@ let defs { defs; _ } =
     | DEF_aux (DEF_mapdef (MD_aux (MD_mapping (i, _, ml), _)), _) ->
         debug_print "DEF_mapdef";
         List.iter (parse_mapcl i) ml
+    | DEF_aux (DEF_register (DEC_aux (DEC_reg (atyp, id, _), _)), _) ->
+        let reg_name = string_of_id id in
+        let reg_type = string_of_typ atyp in
+        Hashtbl.add registers reg_name reg_type
     | _ -> debug_print ~printer:prerr_string ""
   ) defs;
-
```
Comment (ThinkOpenly) [inline src/sail_json_backend/json.ml:605]: lost a blank line here. Add that back, please.

**Review by ThinkOpenly — CHANGES_REQUESTED** [review]
This looks great! One comment on sorting, and a handful of whitespace clean up requested. Thanks!

File: src/sail_json_backend/json.ml (hunk 1 above)
Comment (AbhinavMir) [reply src/sail_json_backend/json.ml]: Oh no! I did not see the whitespaces adding up like that. I squashed a bunch of commits and they must've added up! Sorry, fixing now!

**Comment by AbhinavMir** [comment]
@ThinkOpenly I am running into an odd issue, I am sorting the registers as such

```ocaml
let json_of_registers () =
  let regs = Hashtbl.fold (fun name regtype accum ->
    (name, regtype) :: accum
  ) registers [] in
  let sorted_regs = List.sort (fun (n1, t1) (n2, t2) ->
    match String.compare t1 t2 with
    | 0 ->  (* If types are the same, sort by parsed number *)
       let num1 = parse_reg_name n1 in
       let num2 = parse_reg_name n2 in
       if num1 = num2 then String.compare n1 n2 else compare num1 num2
    | c -> c
  ) regs in
  let json_regs = List.map (fun (name, regtype) ->
    "\n  {\n" ^
    "    \"name\": \"" ^ name ^ "\",\n" ^
    "    \"type\": \"" ^ regtype ^ "\"\n  },"
  ) sorted_regs in
  let reg_list = String.concat "" json_regs in
  let final_reg_list = if reg_list = "" then "" else String.sub reg_list 0 (String.length reg_list - 1) in
  "[ " ^ final_reg_list ^ "\n]"
```

In here, the returned array is x1, x11, x12, x13, x14...., x2, x21..... etc. instead of x1,x2,x3 and so on. Any advice?

**Comment by ThinkOpenly** [comment]
> @ThinkOpenly I am running into an odd issue, I am sorting the registers as such
> ...
> In here, the returned array is x1, x11, x12, x13, x14...., x2, x21..... etc. instead of x1,x2,x3 and so on. Any advice?

Of course. Lexicographical sorting will do that. _sigh_. I think we'll need to live with that. I like your results as-is.

File: src/sail_json_backend/json.ml (hunk 7 above)
Comment (AbhinavMir) [reply src/sail_json_backend/json.ml:605]: Is it added back? I can't tell due to the UI, I see the blank line [here](https://github.com/AbhinavMir/sail/blob/953f3f34bdf7ba9e3ac314822a5a7e6ad0a83aba/src/sail_json_backend/json.ml#L516).

Comment (ThinkOpenly) [reply src/sail_json_backend/json.ml:605]: I see a blank line. The best was to look is to go to the "Files changed" tab.

However, now that I'm looking again, the blank line contains 2 spaces, thus is different than the original line. Remove those 2 spaces, and this should all disappear.

**Comment by ThinkOpenly** [comment]
May I request an update? What do you think still needs to be done here, and are you willing and able to complete that work?

**Comment by AbhinavMir** [comment]
@ThinkOpenly yep! I thought this was closed haha, I think there's just some formatting I wll get to right away

**Comment by AbhinavMir** [comment]
@snapdgn what did you use to format the files? Was it `dune format`?

**Comment by snapdgn** [comment]
> @snapdgn what did you use to format the files? Was it `dune format`?

Yes. `dune fmt` is being used, since upstream also uses it as their default formatter.

**Comment by AbhinavMir** [comment]
Moving to #27

**Pull request closed without merging**